internal/pkg/agent/application/upgrade/watcher.go

```diff
@@ -0,0 +1,217 @@
+						failedTimer.Stop()
+						ch.log.Error("Agent reported healthy (failed timer stopped): %s", err)
+					}
+				}
+				failedErr = err
+				if failedCount > statusFailureFlipFlopsAllowed {
+					err := fmt.Errorf("%w '%d' times in a row", ErrAgentFlipFlopFailed, failedCount)
+					ch.log.Error(err)
+					ch.notifyChan <- err
+				}
+			case <-failedTimer.C:
+				if failedErr == nil {
+					// error was cleared; do nothing
+					continue
+				}
+				// error lasted longer than the checkInterval, notify!
+				ch.notifyChan <- failedErr
+			}
+		}
+	}()
+
+LOOP:
+	for {
+		connectTimer := time.NewTimer(ch.checkInterval)
+		select {
+		case <-ctx.Done():
+			connectTimer.Stop()
+			return
+		case <-connectTimer.C:
+			ch.log.Info("Trying to connect to agent")
+			err := ch.agentClient.Connect(ctx)
+			if err != nil {
+				ch.connectCounter++
+				ch.log.Error("Failed connecting to running daemon: %s", err)
+				if ch.checkFailures() {
+					return
+				}
+				// agent is probably not running
+				continue
+			}
+			watch, err := ch.agentClient.StateWatch(ctx)
+			if err != nil {
+				// considered a connect error
+				ch.agentClient.Disconnect()
+				ch.log.Error("Failed to start state watch: %s", err)
+				ch.connectCounter++
+				if ch.checkFailures() {
+					return
+				}
+				// agent is probably not running
+				continue
+			}
+
+			ch.log.Info("Connected to agent")
+
+			// clear the connectCounter as connection was successfully made
+			// we don't want a disconnect and a reconnect to be counted with
+			// the connectCounter that is tracked with the lostCounter
+			ch.connectCounter = 0
+
+			// failure is tracked only for the life of the connection to
+			// the watch streaming protocol. either an error that last longer
+			// than the checkInterval or to many flopping of error/non-error
+			// will trigger a reported failure
+			failedReset <- true
+			failedCh <- nil
+
+			for {
+				state, err := watch.Recv()
+				if err != nil {
+					// agent has crashed or exited
+					ch.agentClient.Disconnect()
+					ch.log.Error("Failed reading next state: %s", err)
+					ch.lostCounter++
+					if ch.checkFailures() {
+						return
+					}
+					continue LOOP
+				}
+
+				if state.State == client.Failed {
+					// top-level failure (something is really wrong)
+					failedCh <- fmt.Errorf("%w: %s", ErrAgentStatusFailed, state.Message)
+					continue
+				} else {
+					// agent is healthy; but a component might not be healthy
+					// upgrade tracks unhealthy component as an issue with the upgrade
+					var compErr error
+					for _, comp := range state.Components {
+						if comp.State == client.Failed {
+							compErr = multierror.Append(compErr, fmt.Errorf("component %s[%v] failed: %s", comp.Name, comp.ID, comp.Message))
+						}
+					}
+					if compErr != nil {
+						failedCh <- fmt.Errorf("%w: %w", ErrAgentComponentFailed, compErr)
+						continue
+					}
+				}
+
+				// nothing is failed
+				failedCh <- nil
```

Comment:
If nothing is failed, it looks like we continue with this inner loop, immediately calling `watch.Recv()` again, followed by the status checks.  Is this what we want?  Or do we want to continue with `LOOP` here so the next `watch.Recv()` happens when the `connectTimer` fires again?

Reply:
This is a streaming protocol so its a permanent connection that keeps sending state changes. This is correct we want to keep receiving the next state.

Reply:
I see.  So `watch.Recv()` will block until it has something to return?  In that case, we're good here!

Reply:
Correct, it is blocking.